### ISSUE TITLE
About detach_generator_consistency

### ISSUE THREAD

**Issue by JaniceXiong** [Yale-LILY/DYLE#5]
Hi, thanks for your outstanding work!
I saw all the `detach_generator_consistency` in the source code is set to False. In this setting, the `dynamic_mlp` can only be optimized through the `seq_loss `path after marginalizing.

I wonder why we do not set `detach_generator_consistency `to True. Does it because the actual training results were worse? But I did not see it reported in the paper.

If it got worse, maybe the model collapsed because the dynamic scores and doc scores are easily predicted to all zeros or there were other reasons. Could you give me some explanations?
Thanks a lot! @MaoZiming

**Comment by ChenWu98** [comment]
Hi,

Thanks for your interest! ``detach_generator_consistency`` is actually set to True in our code. Can you double-check it?

When ``detach_generator_consistency=True``, the consistency loss is not used to optimize the generator, while it is still used to optimize the extractor (not the ``dynamic_mlp``, sorry about the mistake in the previous reply). This is described in Eq. (5), where the superscript ``eta`` stands for the extractor.

**Comment by JaniceXiong** [comment]
> Hi,
> 
> Thanks for your interest! `detach_generator_consistency` is actually set to True in our code. Can you double-check it?
> 
> When `detach_generator_consistency=True`, the consistency loss is not used to optimize the generator, while it is still used to optimize the `dynamic_mlp`. This is described in Eq. (5), where the superscript `eta` stands for the extractor.

Thanks for your quick reply! Yes, it is set to True. I'm sorry for miswriting it.

But I still feel confused. In the setting of detaching, the consistency loss is only to optimize the extractor because doc_scores are used. Besides, g_probs/dynamic_scores, which are calculated after projecting of dynamic_nlp, are detached from the loss graph. Does this mean that dynamic_nlp is not optimized by the consistency loss because the dynamic_nlp is not connected to the consistency loss in the graph.
```
if config_main.detach_generator_consistency:
   g_probs = g_probs.detach()
r_probs = torch.softmax(doc_scores, dim=1).clamp(min=1e-5)  # shape = (B, n_docs)
consistency_loss = kl_divergence(Categorical(probs=g_probs), Categorical(probs=r_probs)).mean(0)
```
On the other hand, dynamic_nlp can only be optimized by the seq_loss because of dynamic_scores are used in get_nll.

I wonder why we do not set detach_generator_consistency to False and let consistency loss also benefit dynamic_mlp and BART generator? Are there any experimental results indicating that this setting got worse?

**Comment by ChenWu98** [comment]
You are right, the ``dynamic_mlp`` is not optimized by the consistency_loss. Sorry about the mistake in the previous reply. 

Our motivation for consistency loss is to help the extractor learn from the generator. Since ``dynamic_mlp`` learns to assign larger weights to snippets that help generation, we find it beneficial to use the learned weights as a training signal for the extractor. However, the weights from the extractor can be noisy and not necessarily help (even harmful to, shown by early experiments) the generator.
